main/src/library/BPlusTree.flix

```diff
@@ -1,282 +1,488 @@
+    /// Consider the following example where everything is initialized to 0:
+    /// Thread 1 executes
+    /// {a = 2; b = 3}
+    /// Thread 2 executes
+    /// {c = a; d = b}
+    /// The question is what can `c` and `d` be? It depends on the JVM, but can be any of:
+    /// {c = 2, d = 3}, {c = 0, d = 0}, {c = 2, d = 0}, {c = 0, d = 3}
+    /// The last follows from the fact that the JVM can reorder statements under some requirements.
+    /// The requirements are roughly that the new 'code' should be equivalent on a single-threaded machine
+    /// and there are no happens-before relations which link to other threads preventing the reordering.
+    /// If either variable were volatile we might avoid this, but volatile is expensive (and not in Flix).
+    ///
+    /// For us this means that {array[size] = 5; size++} can be reordered to {oldSize = size++; array[oldSize] = 5}.
+    /// Another thread reading may then see the increase of size and try to access that position getting null.
+    /// This has not happened in our experiments, but we wish to handle the case. The risk is in binarySearch and we
+    /// therefore check whether any value read is null and if so 'aborts'.
+    ///
+    /// For Java's memory model see: https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4
+    ///
+
     @Internal
-    pub enum Lock[_: Region](JReentrantLock)
+    pub enum Lock[_: Region](JStampedLock)
 
     mod Lock {
-        import java.util.concurrent.locks.{ReentrantLock => JReentrantLock}
+        import java.util.concurrent.locks.{StampedLock => JStampedLock}
         use BPlusTree.Lock
 
-        /// Unchecked casts are here used to remove the IO effect. It is replaced
-        /// with an effect to a region to ensure that the code is not optimized away.
+        /// Unchecked casts are here used to remove the IO effect. It is replaced with an
+        /// effect to a region to ensure that the code is not optimized away.
+
+        ///
+        /// Wait for `lock` to be unlocked by the current holder of the write-lock on it.
+        ///
+        /// Why:
+        ///
+        /// Virtual threads can show weird behaviour when using all physical threads
+        /// and using locking. Without this the threads using optimistic reading will never be
+        /// removed from the physical thread. Meanwhile a thread that holds the lock will never
+        /// get to run and we have a deadlock. Progress can happen, but it appears to be
+        /// extremely unlikely. To fix this call a blocking operation which makes the JVM remove
+        /// the virtual threads from the physical threads allowing the lock-holder to unlock.
+        ///
+        /// For a better explanation see Netflix's encounter of a similar problem:
+        /// https://netflixtechblog.com/java-21-virtual-threads-dude-wheres-my-lock-3052540e231d?gi=128dba0ad426
+        pub def dumbWait(lock: Lock[r]): Unit \ r =
```

Comment:
I understand that this is hack-ish by necessity, but maybe it should be called `yieldBasedOn` or that describes what it is used for